### DISCUSSION TITLE
Support services in Container `Publish` and `From`

### DISCUSSION THREAD

**Discussion by helderco** [dagger/dagger#12461] [Ideas]
Related to:
- https://github.com/dagger/dagger/issues/5235
- https://github.com/dagger/dagger/issues/6271

There's a workaround using `docker` commands, but having support for `Service` would make it much simpler to manage.

**Comment by helderco** [comment]
@vito, what's the effort here?

**Comment by vito** [comment]
The gist of it is we need to override the built-in Buildkit source (`from`) and exporter (`publish`) with something that starts the dependent service and resolves its address using Dagger's DNS.

For `from` there's lots of prior art here:

* [https://github.com/dagger/dagger/tree/da62b6dae4731f8a7fe9b1373f119a05a044a618/engine/sources](https://github.com/dagger/dagger/tree/da62b6dae4731f8a7fe9b1373f119a05a044a618/engine/sources)
* [https://github.com/dagger/dagger/blob/da62b6dae4731f8a7fe9b1373f119a05a044a618/engine/sources/httpdns/source.go#L107](https://github.com/dagger/dagger/blob/da62b6dae4731f8a7fe9b1373f119a05a044a618/engine/sources/httpdns/source.go#L107)
* [https://github.com/dagger/dagger/blob/da62b6dae4731f8a7fe9b1373f119a05a044a618/engine/sources/netconfhttp/transport.go#L26](https://github.com/dagger/dagger/blob/da62b6dae4731f8a7fe9b1373f119a05a044a618/engine/sources/netconfhttp/transport.go#L26)

We'd basically just copy-paste Buildkit's image source and modify it like we did with these. If there's a better way that involves copying less code go for it - cc @jedevc in case you've got a better idea.

For `publish`, not sure, but seems like we'd be able to figure something out, maybe just calling Buildkit code directly in combination with `netconfhttp`.

**Comment by sagikazarmark** [comment]
This is how I work around it right now:


```go
	image := m.Build().ContainerImage("")

	registry := dag.Registry().Service()

	_, err := dag.Container().From("quay.io/skopeo/stable").
		WithServiceBinding("registry", registry).
		WithMountedFile("/work/image.tar", image.AsTarball()).
		WithExec([]string{"skopeo", "copy", "--all", "--dest-tls-verify=false", "docker-archive:/work/image.tar", "docker://registry:5000/openmeter/openmeter:latest"}).
		Sync(ctx)
	if err != nil {
		return err
	}
```

(Most of it is based on @marcosnils k3s example)

Would be cool to use `Publish` directly.

**Comment by jedevc** [comment]
This will get a lot easier with the work we're doing from dagop/etc.

We should *import* `From` + `Publish` into dagger itself, so we're not calling buildkit here: https://github.com/dagger/dagger/issues/10367

Once we own this logic, we can inject the service info *directly* into the containerd stuff.
